File: connector/src/main/java/org/geysermc/connector/scoreboard/Scoreboard.java
/*
 * Copyright (c) 2019-2021 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.connector.scoreboard;

import com.github.steveice10.mc.protocol.data.game.scoreboard.ScoreboardPosition;
import com.nukkitx.protocol.bedrock.data.ScoreInfo;
import com.nukkitx.protocol.bedrock.packet.RemoveObjectivePacket;
import com.nukkitx.protocol.bedrock.packet.SetDisplayObjectivePacket;
import com.nukkitx.protocol.bedrock.packet.SetScorePacket;
import lombok.Getter;
import org.geysermc.connector.GeyserConnector;
import org.geysermc.connector.GeyserLogger;
import org.geysermc.connector.network.session.GeyserSession;
import org.geysermc.connector.utils.LanguageUtils;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

import static org.geysermc.connector.scoreboard.UpdateType.*;

public final class Scoreboard {
    private final GeyserSession session;
    private final GeyserLogger logger;
    @Getter
    private final AtomicLong nextId = new AtomicLong(0);

    private final Map<String, Objective> objectives = new HashMap<>();
    private final Map<ScoreboardPosition, Objective> objectiveSlots = new HashMap<>();
    private final Map<String, Team> teams = new ConcurrentHashMap<>(); // updated on multiple threads

    private int lastAddScoreCount = 0;
    private int lastRemoveScoreCount = 0;

    public Scoreboard(GeyserSession session) {
        this.session = session;
        this.logger = GeyserConnector.getInstance().getLogger();
    }

    public void removeScoreboard() {
        Iterator<Objective> iterator = objectives.values().iterator();
        while (iterator.hasNext()) {
            Objective objective = iterator.next();
            iterator.remove();
            deleteObjective(objective, false);
        }
    }

    public Objective registerNewObjective(String objectiveId) {
        Objective objective = objectives.get(objectiveId);
        if (objective != null) {
            // we have no other choice, or we have to make a new map?
            // if the objective hasn't been deleted, we have to force it
            if (objective.getUpdateType() != REMOVE) {
                return null;
            }
            deleteObjective(objective, true);
        }

        objective = new Objective(this, objectiveId);
        objectives.put(objectiveId, objective);
        return objective;
    }

    public Objective displayObjective(String objectiveId, ScoreboardPosition displaySlot) {
        Objective objective = objectives.get(objectiveId);
        if (objective == null) {
            return null;
        }

        if (!objective.isActive()) {
            objective.setActive(displaySlot);
            // for reactivated objectives
            objective.setUpdateType(ADD);
        }

        Objective storedObjective = objectiveSlots.get(displaySlot);
        if (storedObjective != null && storedObjective != objective) {
            objective.pendingRemove();
        }
        objectiveSlots.put(displaySlot, objective);
        return objective;
    }

    public Team registerNewTeam(String teamName, Set<String> players) {
        Team team = teams.get(teamName);
        if (team != null) {
            logger.info(LanguageUtils.getLocaleStringLog("geyser.network.translator.team.failed_overrides", teamName));
            return team;
        }

        team = new Team(this, teamName).addEntities(players);
        teams.put(teamName, team);
        return team;
    }

    public Objective getObjective(String objectiveName) {
        return objectives.get(objectiveName);
    }

    public Collection<Objective> getObjectives() {
        return objectives.values();
    }

    public Team getTeam(String teamName) {
        return teams.get(teamName);
    }

    public void unregisterObjective(String objectiveName) {
        Objective objective = getObjective(objectiveName);
        if (objective != null) {
            objective.pendingRemove();
        }
    }

    public void removeTeam(String teamName) {
        Team remove = teams.remove(teamName);
        if (remove != null) {
            remove.setUpdateType(REMOVE);
        }
    }

    public void onUpdate() {
        List<ScoreInfo> addScores = new ArrayList<>(lastAddScoreCount);
        List<ScoreInfo> removeScores = new ArrayList<>(lastRemoveScoreCount);
        List<Objective> removedObjectives = new ArrayList<>();

        Team playerTeam = getTeamFor(session.getPlayerEntity().getUsername());
        Objective correctSidebar = null;

        for (Objective objective : objectives.values()) {
            // objective has been deleted
            if (objective.getUpdateType() == REMOVE) {
                removedObjectives.add(objective);
                continue;
            }

            if (playerTeam != null && playerTeam.getColor() == objective.getTeamColor()) {
                correctSidebar = objective;
            }
        }

        if (correctSidebar == null) {
            correctSidebar = objectiveSlots.get(ScoreboardPosition.SIDEBAR);
        }

        handleObjective(objectiveSlots.get(ScoreboardPosition.PLAYER_LIST), addScores, removeScores);
        handleObjective(correctSidebar, addScores, removeScores);
        handleObjective(objectiveSlots.get(ScoreboardPosition.BELOW_NAME), addScores, removeScores);

        Iterator<Team> teamIterator = teams.values().iterator();
        while (teamIterator.hasNext()) {
            Team current = teamIterator.next();

            switch (current.getCachedUpdateType()) {
                case ADD:
                case UPDATE:
                    current.markUpdated();
                    break;
                case REMOVE:
                    teamIterator.remove();
            }
        }

        if (!removeScores.isEmpty()) {
            SetScorePacket setScorePacket = new SetScorePacket();
            setScorePacket.setAction(SetScorePacket.Action.REMOVE);
            setScorePacket.setInfos(removeScores);
            session.sendUpstreamPacket(setScorePacket);
        }

        if (!addScores.isEmpty()) {
            SetScorePacket setScorePacket = new SetScorePacket();
            setScorePacket.setAction(SetScorePacket.Action.SET);
            setScorePacket.setInfos(addScores);
            session.sendUpstreamPacket(setScorePacket);
        }

        for (Objective objective : removedObjectives) {
            deleteObjective(objective, true);
        }

        lastAddScoreCount = addScores.size();
        lastRemoveScoreCount = removeScores.size();
    }

    private void handleObjective(Objective objective, List<ScoreInfo> addScores, List<ScoreInfo> removeScores) {
        if (objective == null || objective.getUpdateType() == REMOVE) {
            return;
        }

        // hearts can't hold teams, so we treat them differently
        if (objective.getType() == 1) {
            for (Score score : objective.getScores().values()) {
                boolean update = score.shouldUpdate();

                if (update) {
                    score.update(objective.getObjectiveName());
                }

                if (score.getUpdateType() != REMOVE && update) {
                    addScores.add(score.getCachedInfo());
                }
                if (score.getUpdateType() != ADD && update) {
                    removeScores.add(score.getCachedInfo());
                }
            }
            return;
        }

        boolean objectiveUpdate = objective.getUpdateType() == UPDATE;
        boolean objectiveAdd = objective.getUpdateType() == ADD;

        for (Score score : objective.getScores().values()) {
            Team team = score.getTeam();

            boolean add = objectiveAdd || objectiveUpdate;
            boolean remove = false;
            if (team != null) {
                if (team.getUpdateType() == REMOVE || !team.hasEntity(score.getName())) {
                    score.setTeam(null);
                    add = true;
                    remove = true;
                }
            }

            add |= score.shouldUpdate();
            remove |= score.shouldUpdate();

            if (score.getUpdateType() == REMOVE) {
                add = false;
            }

            if (score.getUpdateType() == ADD) {
                remove = false;
            }

            if (score.shouldUpdate()) {
                score.update(objective.getObjectiveName());
            }

            if (add) {
                addScores.add(score.getCachedInfo());
            }

            if (remove) {
                removeScores.add(score.getCachedInfo());
            }

            // score is pending to be removed, so we can remove it from the objective
            if (score.getUpdateType() == REMOVE) {
                objective.removeScore0(score.getName());
            }

            score.setUpdateType(NOTHING);
        }

        if (objectiveUpdate) {
            RemoveObjectivePacket removeObjectivePacket = new RemoveObjectivePacket();
            removeObjectivePacket.setObjectiveId(objective.getObjectiveName());
            session.sendUpstreamPacket(removeObjectivePacket);
        }

        if (objectiveAdd || objectiveUpdate) {
            SetDisplayObjectivePacket displayObjectivePacket = new SetDisplayObjectivePacket();
            displayObjectivePacket.setObjectiveId(objective.getObjectiveName());
            displayObjectivePacket.setDisplayName(objective.getDisplayName());
            displayObjectivePacket.setCriteria("dummy");
            displayObjectivePacket.setDisplaySlot(objective.getDisplaySlotName());
            displayObjectivePacket.setSortOrder(1); // 0 = ascending, 1 = descending
            session.sendUpstreamPacket(displayObjectivePacket);
        }

        objective.setUpdateType(NOTHING);
    }

    /**
     * @param remove if we should remove the objective from the objectives map.
     */
    public void deleteObjective(Objective objective, boolean remove) {
        if (remove) {
            objectives.remove(objective.getObjectiveName());
        }
        objectiveSlots.remove(objective.getDisplaySlot(), objective);

        objective.removed();

        RemoveObjectivePacket removeObjectivePacket = new RemoveObjectivePacket();
        removeObjectivePacket.setObjectiveId(objective.getObjectiveName());
        session.sendUpstreamPacket(removeObjectivePacket);
    }

    public Team getTeamFor(String entity) {
        for (Team team : teams.values()) {
            if (team.hasEntity(entity)) {
                return team;
            }
        }
        return null;
    }
}

File: core/src/main/java/org/geysermc/geyser/scoreboard/Scoreboard.cpp
/*
 * Copyright (c) 2019-2022 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.geyser.scoreboard;

#include "it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap"
#include "lombok.Getter"
#include "net.kyori.adventure.text.Component"
#include "org.checkerframework.checker.nullness.qual.Nullable"
#include "org.cloudburstmc.protocol.bedrock.data.ScoreInfo"
#include "org.cloudburstmc.protocol.bedrock.data.command.CommandEnumConstraint"
#include "org.cloudburstmc.protocol.bedrock.packet.SetScorePacket"
#include "org.geysermc.geyser.GeyserImpl"
#include "org.geysermc.geyser.GeyserLogger"
#include "org.geysermc.geyser.entity.type.Entity"
#include "org.geysermc.geyser.entity.type.player.PlayerEntity"
#include "org.geysermc.geyser.scoreboard.display.slot.BelownameDisplaySlot"
#include "org.geysermc.geyser.scoreboard.display.slot.DisplaySlot"
#include "org.geysermc.geyser.scoreboard.display.slot.PlayerlistDisplaySlot"
#include "org.geysermc.geyser.scoreboard.display.slot.SidebarDisplaySlot"
#include "org.geysermc.geyser.session.GeyserSession"
#include "org.geysermc.mcprotocollib.protocol.data.game.scoreboard.NameTagVisibility"
#include "org.geysermc.mcprotocollib.protocol.data.game.scoreboard.ScoreboardPosition"
#include "org.geysermc.mcprotocollib.protocol.data.game.scoreboard.TeamColor"
#include "org.jetbrains.annotations.Contract"

#include "java.util.ArrayList"
#include "java.util.Collections"
#include "java.util.EnumMap"
#include "java.util.EnumSet"
#include "java.util.HashMap"
#include "java.util.LinkedHashMap"
#include "java.util.List"
#include "java.util.Map"
#include "java.util.Set"
#include "java.util.concurrent.ConcurrentHashMap"
#include "java.util.concurrent.atomic.AtomicBoolean"
#include "java.util.concurrent.atomic.AtomicLong"
#include "java.util.function.Function"
#include "java.util.stream.Collectors"

#include "static org.geysermc.geyser.scoreboard.UpdateType.REMOVE"


public final class Scoreboard {
    private static final bool SHOW_SCOREBOARD_LOGS = Boolean.parseBoolean(System.getProperty("Geyser.ShowScoreboardLogs", "true"));
    private static final bool ADD_TEAM_SUGGESTIONS = Boolean.parseBoolean(
        System.getProperty("Geyser.AddTeamSuggestions", std::string.valueOf(GeyserImpl.getInstance().config().advanced().addTeamSuggestions()))
    );

    private final GeyserSession session;
    private final GeyserLogger logger;
    private final AtomicLong nextId = new AtomicLong(0);

    private final Map<std::string, Objective> objectives = new ConcurrentHashMap<>();
    @Getter
    private final Map<ScoreboardPosition, DisplaySlot> objectiveSlots = Collections.synchronizedMap(new EnumMap<>(ScoreboardPosition.class));
    private final List<DisplaySlot> removedSlots = Collections.synchronizedList(new ArrayList<>());

    private final Map<std::string, Team> teams = new ConcurrentHashMap<>();

    @Getter
    private final Map<std::string, Team> playerToTeam = new Object2ObjectOpenHashMap<>();

    private final AtomicBoolean updateLockActive = new AtomicBoolean(false);
    private int lastAddScoreCount = 0;
    private int lastRemoveScoreCount = 0;

    public Scoreboard(GeyserSession session) {
        this.session = session;
        this.logger = GeyserImpl.getInstance().getLogger();
    }

    public void removeScoreboard() {
        var copy = new HashMap<>(objectiveSlots);
        objectiveSlots.clear();

        for (DisplaySlot slot : copy.values()) {
            slot.remove();
        }
    }

    public Objective registerNewObjective(std::string objectiveId) {
        Objective objective = objectives.get(objectiveId);
        if (objective != null) {

            if (SHOW_SCOREBOARD_LOGS) {
                logger.warning("An objective with the same name '" + objectiveId + "' already exists! Ignoring new objective!");
            }
            return null;
        }

        objective = new Objective(this, objectiveId);
        objectives.put(objectiveId, objective);
        return objective;
    }

    public void displayObjective(std::string objectiveId, ScoreboardPosition slot) {
        if (objectiveId.isEmpty()) {

            var display = objectiveSlots.get(slot);
            if (display != null) {
                removedSlots.add(display);
                objectiveSlots.remove(slot, display);
                var objective = display.objective();
                objective.removeDisplaySlot(display);
            }
            return;
        }

        Objective objective = objectives.get(objectiveId);
        if (objective == null) {
            return;
        }

        var display = objectiveSlots.get(slot);
        if (display != null && display.objective() != objective) {
            removedSlots.add(display);
        }

        display = switch (DisplaySlot.slotCategory(slot)) {
            case SIDEBAR -> new SidebarDisplaySlot(session, objective, slot);
            case BELOW_NAME -> new BelownameDisplaySlot(session, objective);
            case PLAYER_LIST -> new PlayerlistDisplaySlot(session, objective);
            default -> throw new IllegalStateException("Unexpected value: " + slot);
        };
        objectiveSlots.put(slot, display);
        objective.addDisplaySlot(display);
    }

    public void registerNewTeam(
        std::string teamName,
        String[] players,
        Component name,
        Component prefix,
        Component suffix,
        NameTagVisibility visibility,
        TeamColor color
    ) {
        Team team = teams.get(teamName);
        if (team != null) {
            if (SHOW_SCOREBOARD_LOGS) {
                logger.info("Ignoring team %s for %s. It overrides without removing old team.".formatted(teamName, session.javaUsername()));
            }
            return;
        }

        team = new Team(this, teamName, players, name, prefix, suffix, visibility, color);
        teams.put(teamName, team);


        if (ADD_TEAM_SUGGESTIONS) {
            session.addCommandEnum("Geyser_Teams", team.id());
        }
    }

    public void onUpdate() {

        if (updateLockActive.getAndSet(true)) {
            return;
        }

        List<ScoreInfo> addScores = new ArrayList<>(lastAddScoreCount);
        List<ScoreInfo> removeScores = new ArrayList<>(lastRemoveScoreCount);

        Team playerTeam = getTeamFor(session.getPlayerEntity().getUsername());
        DisplaySlot correctSidebarSlot = null;

        for (DisplaySlot slot : objectiveSlots.values()) {

            if (slot.updateType() == REMOVE) {
                continue;
            }

            if (playerTeam != null && playerTeam.color() == slot.teamColor()) {
                correctSidebarSlot = slot;
            }
        }

        if (correctSidebarSlot == null) {
            correctSidebarSlot = objectiveSlots.get(ScoreboardPosition.SIDEBAR);
        }

        var actualRemovedSlots = new ArrayList<>(removedSlots);
        for (var slot : actualRemovedSlots) {


            slot.remove();
        }
        removedSlots.removeAll(actualRemovedSlots);

        handleDisplaySlot(objectiveSlots.get(ScoreboardPosition.PLAYER_LIST), addScores, removeScores);
        handleDisplaySlot(correctSidebarSlot, addScores, removeScores);
        handleDisplaySlot(objectiveSlots.get(ScoreboardPosition.BELOW_NAME), addScores, removeScores);

        if (!removeScores.isEmpty()) {
            SetScorePacket packet = new SetScorePacket();
            packet.setAction(SetScorePacket.Action.REMOVE);
            packet.setInfos(removeScores);
            session.sendUpstreamPacket(packet);
        }

        if (!addScores.isEmpty()) {
            SetScorePacket packet = new SetScorePacket();
            packet.setAction(SetScorePacket.Action.SET);
            packet.setInfos(addScores);
            session.sendUpstreamPacket(packet);
        }

        lastAddScoreCount = addScores.size();
        lastRemoveScoreCount = removeScores.size();
        updateLockActive.set(false);
    }

    private void handleDisplaySlot(DisplaySlot slot, List<ScoreInfo> addScores, List<ScoreInfo> removeScores) {
        if (slot != null) {
            slot.render(addScores, removeScores);
        }
    }

    public Objective getObjective(std::string objectiveName) {
        return objectives.get(objectiveName);
    }

    public void removeObjective(Objective objective) {
        objectives.remove(objective.getObjectiveName());
        for (DisplaySlot slot : objective.getActiveSlots()) {
            objectiveSlots.remove(slot.position(), slot);
            removedSlots.add(slot);
        }
    }

    public void resetPlayerScores(std::string playerNameOrEntityUuid) {
        for (Objective objective : objectives.values()) {
            objective.removeScore(playerNameOrEntityUuid);
        }
    }

    public Team getTeam(std::string teamName) {
        return teams.get(teamName);
    }

    public Team getTeamFor(std::string playerNameOrEntityUuid) {
        return playerToTeam.get(playerNameOrEntityUuid);
    }

    public void removeTeam(std::string teamName) {
        Team remove = teams.remove(teamName);
        if (remove == null) {
            return;
        }
        remove.remove();
        session.removeCommandEnum("Geyser_Teams", remove.id());
    }

    @Contract("-> new")
    public Map<std::string, Set<CommandEnumConstraint>> getTeamNames() {
        return teams.keySet().stream()
                .collect(Collectors.toMap(Function.identity(), o -> EnumSet.noneOf(CommandEnumConstraint.class),
                        (o1, o2) -> o1, LinkedHashMap::new));
    }

    public void playerRegistered(PlayerEntity player) {
        for (DisplaySlot slot : objectiveSlots.values()) {
            slot.playerRegistered(player);
        }
    }

    public void playerRemoved(PlayerEntity player) {
        for (DisplaySlot slot : objectiveSlots.values()) {
            slot.playerRemoved(player);
        }
    }

    public void entityRegistered(Entity entity) {
        var team = getTeamFor(entity.teamIdentifier());
        if (team != null) {
            team.onEntitySpawn(entity);
        }
    }

    public void entityRemoved(Entity entity) {
        var team = getTeamFor(entity.teamIdentifier());
        if (team != null) {
            team.onEntityRemove(entity);
        }
    }

    public void setTeamFor(Team team, Set<std::string> entities) {
        for (DisplaySlot slot : objectiveSlots.values()) {

            if (slot instanceof SidebarDisplaySlot sidebar) {
                sidebar.setTeamFor(team, entities);
            }
        }
    }

    public long nextId() {
        return nextId.getAndIncrement();
    }

    public GeyserSession session() {
        return session;
    }
}
